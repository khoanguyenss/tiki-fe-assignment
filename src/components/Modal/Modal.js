import React from 'react';
import PropTypes from 'prop-types';
import { Wrapper, HomeLink, RestartBtn } from './Modal.styled';

const statusText = ['won', 'lost'];

const Modal = ({
  isOpen, duration, fetchBoard, gameStatus,
}) => (
  <Wrapper isOpen={isOpen}>
    <div>{`You ${statusText[gameStatus - 2]} the game in ${duration}`}</div>
    <RestartBtn onClick={() => fetchBoard()}>New Game</RestartBtn>
    <HomeLink to="/">Home Page</HomeLink>
  </Wrapper>
);

Modal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  duration: PropTypes.string.isRequired,
  fetchBoard: PropTypes.func.isRequired,
  gameStatus: PropTypes.number.isRequired,
};

export default Modal;
