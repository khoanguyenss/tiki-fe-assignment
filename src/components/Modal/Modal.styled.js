import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';

export const Wrapper = styled.div`
  display: none;
  position: fixed;
  width: 100vw;
  height: 100vh;
  overflow: hidden;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: rgba(0, 0, 0, 0.9);
  top: 0px;
  left: 0px;
  color: #fff;
  font-size: 24px;
  
  ${(props) => props.isOpen && css`
    display: flex;
  `}
`;

export const RestartBtn = styled.button`
  font-size: 24px;
  margin-top: 30px;
  margin-bottom: 15px;
  background: unset;
  border: unset;
  outline: unset;
  color: #fff;
  
  &:hover {
    cursor:pointer;
    text-decoration: underline;
  }
`;

export const HomeLink = styled(Link)`
  color: #fff;
  text-decoration: none;
  
  &:hover {
    cursor:pointer;
    text-decoration: underline;
  }
`;
