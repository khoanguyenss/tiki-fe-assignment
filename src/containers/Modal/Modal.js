import React from 'react';
import { useLocation } from 'react-router';
import moment from 'moment';
import Modal from '../../components/Modal';
import {
  actions as MineSweeperActions,
  selectors as MineSweeperSelectors,
} from '../../reducers/minesweeper';
import { useActions, useShallowEqualSelector } from '../../utils/hooks';

const ModalContainer = () => {
  const [
    fetchMineSweeperBeginnerBoard,
    fetchMineSweeperAdvantageBoard,
  ] = useActions([
    MineSweeperActions.fetchMineSweeperBeginnerBoard,
    MineSweeperActions.fetchMineSweeperAdvantageBoard,
  ]);
  const { pathname } = useLocation();
  let fetchBoard = () => null;
  const gameStatus = useShallowEqualSelector(MineSweeperSelectors.getGameStatus);
  const startTime = useShallowEqualSelector(MineSweeperSelectors.getTimer);
  const diff = startTime ? moment().diff(startTime, 'millisecond') : 0;
  const duration = moment(diff).utc().format('HH:mm:ss');

  if (pathname.includes('begin')) fetchBoard = fetchMineSweeperBeginnerBoard;
  else fetchBoard = fetchMineSweeperAdvantageBoard;

  return (
    <Modal
      isOpen={gameStatus > 1}
      gameStatus={gameStatus}
      fetchBoard={fetchBoard}
      duration={duration}
    />
  );
};

export default ModalContainer;
