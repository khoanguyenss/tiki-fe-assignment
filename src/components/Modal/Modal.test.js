import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter as Router, Link } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import Modal from './Modal';
import { RestartBtn } from './Modal.styled';


describe('<Modal />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const fetchBoardMock = jest.fn();

  const Component = mount(
    <Router>
      <Modal
        isOpen
        duration="00:00:00"
        gameStatus={2}
        fetchBoard={fetchBoardMock}
      />
    </Router>
    ,
  );

  it('should render', () => {
    expect(Component.length).toBe(1);
  });
  it('match snapshot', () => {
    expect(Component).toMatchSnapshot();
  });
  it('should trigger fetchBoardMock when Restart button is clicked', () => {
    const RestartBtnComp = Component.find(RestartBtn).at(0);
    RestartBtnComp.simulate('click');
    expect(fetchBoardMock).toBeCalledTimes(1);
  });
});
