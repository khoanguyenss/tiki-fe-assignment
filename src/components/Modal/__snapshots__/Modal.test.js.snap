// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> match snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Modal
      duration="00:00:00"
      fetchBoard={[MockFunction]}
      gameStatus={2}
      isOpen={true}
    >
      <styled.div
        isOpen={true}
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "cNSmUR",
                "rules": Array [
                  "
  display: none;
  position: fixed;
  width: 100vw;
  height: 100vh;
  overflow: hidden;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: rgba(0, 0, 0, 0.9);
  top: 0px;
  left: 0px;
  color: #fff;
  font-size: 24px;
  
  ",
                  [Function],
                  "
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          isOpen={true}
        >
          <div
            className="sc-bdVaJa cNSmUR"
          >
            <div>
              You won the game in 00:00:00
            </div>
            <styled.button
              onClick={[Function]}
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": false,
                      "lastClassName": "dqivxj",
                      "rules": Array [
                        "
  font-size: 24px;
  margin-top: 30px;
  margin-bottom: 15px;
  background: unset;
  border: unset;
  outline: unset;
  color: #fff;
  
  &:hover {
    cursor:pointer;
    text-decoration: underline;
  }
",
                      ],
                    },
                    "displayName": "styled.button",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "button",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                onClick={[Function]}
              >
                <button
                  className="sc-bwzfXH dqivxj"
                  onClick={[Function]}
                >
                  New Game
                </button>
              </StyledComponent>
            </styled.button>
            <Styled(Link)
              to="/"
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-htpNat",
                      "isStatic": false,
                      "lastClassName": "khMNoN",
                      "rules": Array [
                        "
  color: #fff;
  text-decoration: none;
  
  &:hover {
    cursor:pointer;
    text-decoration: underline;
  }
",
                      ],
                    },
                    "displayName": "Styled(Link)",
                    "foldedComponentIds": Array [],
                    "propTypes": Object {
                      "innerRef": [Function],
                      "onClick": [Function],
                      "replace": [Function],
                      "target": [Function],
                      "to": [Function],
                    },
                    "render": [Function],
                    "styledComponentId": "sc-htpNat",
                    "target": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "displayName": "Link",
                      "propTypes": Object {
                        "innerRef": [Function],
                        "onClick": [Function],
                        "replace": [Function],
                        "target": [Function],
                        "to": [Function],
                      },
                      "render": [Function],
                    },
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                to="/"
              >
                <Link
                  className="sc-htpNat khMNoN"
                  to="/"
                >
                  <LinkAnchor
                    className="sc-htpNat khMNoN"
                    href="/"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-htpNat khMNoN"
                      href="/"
                      onClick={[Function]}
                    >
                      Home Page
                    </a>
                  </LinkAnchor>
                </Link>
              </StyledComponent>
            </Styled(Link)>
          </div>
        </StyledComponent>
      </styled.div>
    </Modal>
  </Router>
</BrowserRouter>
`;
